### PR TITLE
Deprecate the internal switching of a backpropagation device

### PR DESCRIPTION
**Pull request by albi3ro**
**Context:**

The internal switching of one device for another backpropagation-compatible device was developed to allow the substitution of `default.qubit.legacy` with either `default.qubit.autograd`, `default.qubit.jax`, `default.qubit.torch`, or `default.qubit.tf`.  This logic has caused a lot of pain over the years, and all five of those devices are now deprecated.  In order to complete the removal of the 5 legacy devices, we need to also deprecate and remove the backprop device switching logic.

Otherwise, it would be very difficult and painful to test the switching of a backpropagation device.

**Description of the Change:**

Deprecates the switching of a device with a different device.  A warning is only raised when the device is not `default.qubit.legacy`, as default.qubit.legacy raises it's own deprecation warnings.

**Benefits:**

We can finally clean up QNode 🎉 💥 

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -132,6 +132,10 @@
   Instead, use `default.qubit` as it now supports backpropagation through the several backends.
   [(#5997)](https://github.com/PennyLaneAI/pennylane/pull/5997)
 
+* The logic for internally switching a device for a different backpropagation
+  compatible device is now deprecated, as it was in place for the deprecated
+  `default.qubit.legacy`.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:137]:
```suggestion
  `default.qubit.legacy`.
  [(#6032)](https://github.com/PennyLaneAI/pennylane/pull/6032)
```

File: pennylane/workflow/qnode.py [hunk 2]
```diff
@@ -878,6 +878,14 @@ def _validate_backprop_method(device, interface, tape=None):
                 if backprop_devices[mapped_interface] == device.short_name:
                     return "backprop", {}, device
 
+                if device.short_name != "default.qubit.legacy":
+                    warnings.warn(
+                        "The switching of devices for backpropagation is now deprecated in v0.38 and "
+                        "will be removed in v0.39, as this behavior was developered purely for the"
```
Comment (Shiro-Raven) [inline pennylane/workflow/qnode.py:884]:
```suggestion
                        "will be removed in v0.39, as this behavior was developed purely for the"
```

**Review by Shiro-Raven — APPROVED** [review]

File: tests/resource/test_specs.py [hunk 3]
```diff
@@ -29,6 +29,13 @@
         (qml.device("default.qubit.legacy", wires=2), 2),
     ]
 
+with pytest.warns(qml.PennyLaneDeprecationWarning):
+    devices_list = [
+        (qml.device("default.qubit"), 1),
+        (qml.device("default.qubit", wires=2), 2),
+        (qml.device("default.qubit.legacy", wires=2), 2),
+    ]
+
```
Comment (albi3ro) [inline tests/resource/test_specs.py:38]:
```suggestion
```

File: pennylane/workflow/qnode.py [hunk 4]
```diff
@@ -878,6 +878,14 @@ def _validate_backprop_method(device, interface, tape=None):
                 if backprop_devices[mapped_interface] == device.short_name:
                     return "backprop", {}, device
 
+                if device.short_name != "default.qubit.legacy":
+                    warnings.warn(
+                        "The switching of devices for backpropagation is now deprecated in v0.38 and "
+                        "will be removed in v0.39, as this behavior was developed purely for the"
+                        " deprecated default.qubit.legacy. ",
```
Comment (mudit2812) [inline pennylane/workflow/qnode.py:885]:
```suggestion
                        "will be removed in v0.39, as this behavior was developed purely for the "
                        "deprecated default.qubit.legacy. ",

```

**Review by mudit2812 — APPROVED** [review]

File: pennylane/workflow/qnode.py [hunk 5]
```diff
@@ -878,6 +878,14 @@ def _validate_backprop_method(device, interface, tape=None):
                 if backprop_devices[mapped_interface] == device.short_name:
                     return "backprop", {}, device
 
+                if device.short_name != "default.qubit.legacy":
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/workflow/qnode.py:881]: Exactly. DQL and it's child devices are all already deprecated.

**Pull request merged**